CRM/Queue/Service.php

```diff
@@ -104,22 +109,55 @@ public function create($queueSpec) {
     return $queue;
   }
 
+  /**
+   * Find/create the queue-spec. Specifically:
+   *
+   * - If there is a stored queue, use its spec.
+   * - If there is no stored queue, and if we have enough information, then create queue.
+   *
+   * @param array $queueSpec
+   * @return array
+   *   Updated queueSpec.
+   * @throws \CRM_Core_Exception
+   */
+  protected function findCreateQueueSpec(array $queueSpec): array {
+    $storageFields = ['type', 'is_autorun'];
+    $dao = new CRM_Queue_DAO_Queue();
+    $dao->name = $queueSpec['name'];
+    if ($dao->find(TRUE)) {
+      return array_merge($queueSpec, CRM_Utils_Array::subset((array) $dao, $storageFields));
```

Comment:
Oh, I forgot that `$dao` had its own `toArray()`. Sounds a bit better than `(array)$dao`. I don't think it's consequential in this formulation, but I'll switch it because it.

Regarding the field-merging process... I agree that (*in the undesirable case of a conflict*) the stored `type` should take precedence. I believe that's what it does already. (I'll double-check before pushing the next rebase.)

It might be helpful to verbalize the situation a bit more:

* (a) there's schema for `CRM_Queue_Service`'s `$queueSpec`,
* (b) there's schema for `civicrm_queue`,
* (c) the schemas overlap (`name`, `type`, `is_autorun` are the same),
* but (d) they do not overlap perfectly (eg `reset` and `is_persistent` are valid in `CRM_Queue_Service` but unintelligible for `civicrm_queue`; `id` is supported by `civicrm_queue` but not `CRM_Queue_Service`).

The `array_merge()` is trying to update `$queueSpec` with overrides/fill-ins from `$dao`. But you only want overlapping fields. Breaking it down:

* `CRM_Utils_Array::subset((array) $dao, $storageFields)`: Take the `$dao` and extract common fields.
* `array_merge($queueSpec, ...$dao...)`: Start with the requested values, then fill-in (override) with DAO values.

Tangentially... I don't have an intuition for how `array_intersect_{key,ukey,assoc,uassoc}` work. I do find PHP's distinction between `+` and `array_merge()` slightly confusing, but I've got more intuition for `array_merge()`. To my recollection(+testing), precedence works like this:

```php
$x = array_merge($raw, $overrides); // $overrides take precedence
$x = $overrides + $raw;             // $overrides take precedence
$x = array_merge($defaults, $raw);  // $raw takes precedence
$x = $raw + defaults;               // $raw takes precedence
```

> * look up by name
> * if found, return the persisted data for that queue (which may differ from queue spec.)
> * if not found, add defaults, create queue, return the values.

1. ✅ Does that.
2. That would drop keys like `reset` which are relevant to `$queueSpec` but not relevant to `civicrm_queue`. 
3. ✅ Does that. Note: it seemed arbitrary to me whether you add defaults to that `$dao` or make a new `$dao`.

Reply:
Updated with `toArray()`. Confirmed that it gives precedence to stored values. Sqaushed.